### DISCUSSION TITLE
Conditional rendering of some sections based on python conditions

### DISCUSSION THREAD

**Discussion by Alternativeheat** [quarto-dev/quarto-cli#9721] [Q&A]
### Description

I am reading a directory containing dataframes which may contain null values for a particular column (let us name it column _**ABC**_). 

I would like to render some sections only if column _**ABC**_ has no missing values. 

````qmd
---
title: "Results"
title-slide-attributes:
    data-background-color: "#e5f5f9"
format:
  revealjs:
    slide-number: true
jupyter: python3
---

```{python}
import pandas as pd

# dataframe list
dataframe_list = [df1, df2, df3]

for i, df in enumerate(dataframe_list):
    null_values = df['ABC'].isnull.any()

    if null_values:
        # execute section A only
    else: 
        # execute both section A and section B
```

<!-- Section A should always be rendered regardless of what values column ABC contains-->
## Section A 
```{python}
print("Column A may or may not contain null values.")
```

<!-- Section B should only be rendered if column ABC has no missing values  -->
## Section B
```{python}
print("Column A has no null values")
```
````

**Comment by mcanouil** [comment]
Add an if statement in your code cell and output markdown instead of print via `IPython`.

**Reply by Alternativeheat** [reply]
Can you be more descriptive, please? A newbie here!

**Reply by mcanouil** [reply]
It's really an if statement and the use of `IPython` module (https://ipython.readthedocs.io/en/stable/).
There is nothing fancy or specific to Quarto here.

````qmd
---
title: "Reproducible Quarto Document"
format: html
---

```{python}
from IPython.display import display, Markdown

A = 0
if A == 1:
  display(Markdown("# Section A"))
```

```{python}
A = 1
if A == 1:
  display(Markdown("# Section B"))
```

````

**Reply by Alternativeheat** [reply]
I am sorry I should have made it clear in the question itself. Tried to make the question simple to understand the essence of how conditional rendering works. But what I want to achieve is more complex than what I defined in my question. 

The end goal is to create **individual reveal.js presentations for individual dataframes** with some content. The content of the presentation will differ for the dataframes based on whether null values are present in the column 'ABC' or not. 

**_The code provided by you is still not giving me the result that I am looking for._** 


````qmd
---
title: "Reproducible Quarto Document"
format: revealjs
---

```{python}
#| echo: false
from IPython.display import display, Markdown

A = 0
if A == 1:
    Markdown("## Section A")
else:
    Markdown("## Section B")
```

## Section A 

One

## Section B

Zero
```` 


_In my output, both section A and section B are rendered rather than just section A as expected._

**What do you think is missing in the code? Thanks!**

**Reply by mcanouil** [reply]
You hardcoded the sections in your source file, thus you get them in the output.
You also removed the display call, meaning your code cell is useless here.

My code shows the logic to display or not content based on computed elements.
It's up to you to use that logic to make what you want.

There are no other choices (right now) for conditional content based on computed elements.

**Reply by Alternativeheat** [reply]
Ah okay, I get what you mean! 

But, can we do more than just displaying the texts (for eg: further processing of the dataframe if the condition is satisfied)? 

In my case, I would like to use the 'ABC' column to create maps and other statistics to display in the revealjs presentation. 

Is this possible with the current state of Quarto? 

````qmd
---
title: "Reproducible Quarto Document"
format: revealjs
---

```{python}
import pandas as pd

# dataframe list
dataframe_list = [df1, df2, df3]

for i, df in enumerate(dataframe_list):
    null_values = df['ABC'].isnull.any()

    if null_values:
        display(Markdown("Null values"))
        # calculates number of rows and columns, find min,max and median of specific columns and displays them to users
        # and more calculations.....
        # Ideally, would like to have a section (## NULL VALUES) in revealjs presentation that displays the results from the calculations. 

    else: 
        display(Markdown("No null values"))
        # use the 'ABC' column to calculate the centroid of the dataframe
        # make map(s) using 'ABC' and other columns
        # and more calculations.....
       # Ideally, would like to have a section (## NO NULL VALUES) in revealjs presentation that displays the map and other results from the calculations. 

```

````

**Reply by cderv** [reply]
@Alternativeheat you can do anything that Python allows you to do as computation, and as long a you output Markdown from you code cells, it should render fine in Quarto. You can add `keep-md: true` in YAML to check what is the content you are creating in your cells. 

You should try and see what comes up for your tries. 

There are more feature you can play with in Quarto like conditional content markdown block (https://quarto.org/docs/authoring/conditional.html) or including external .qmd (https://quarto.org/docs/authoring/includes.html). 
Maybe others can be useful so that you can come up with a workflow that do what you want. 

Quarto may not specifically provide a built in feature for you specific need, but you can probably script around it.

Hope it helps. 

Thanks !
